Add a reader for EPS-SG Ice Cloud Imager

BengtRydberg opened the pull request:
<!-- Describe what your PR does, and why -->
Adding a level1b reader for EPS-SG Ice Cloud Imager (ICI) that will be one the instrument on metop-SGB and will be launched
around 2025. Such a reader is presently not available here, and the aim is to use satpy within ICI level2 processing chain. 

The reader is configured to be able to read some of the datasets within the file, but not all. 

The ICI Level 1B Product Format Specification V3A: https://www.eumetsat.int/media/47582
Eumetsat has published some test data here:  https://www.eumetsat.int/release-simulated-eps-sg-mwi-ici-test-data-2021



<!-- For works in progress choose "Create draft pull request" from the drop-down green button. -->

 - [ ] Closes #xxxx (there is no related issue, should I add one?)<!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [x] Tests added <!-- for all bug fixes or enhancements -->
 - [ ] Fully documented (not 100% sure what is expected)<!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->
 - [x] Add your name to `AUTHORS.md` if not there already

mraspaud commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -0,0 +1,441 @@
+        horns = longitude.n_horns
+        n_scan = longitude.n_scan
+        n_subs = longitude.n_subs
+        lons = np.zeros((n_scan.size, n_samples, horns.size))
+        lats = np.zeros((n_scan.size, n_samples, horns.size))
+        n_subs = np.linspace(0, n_samples - 1, n_subs.size).astype(int)
+        for horn in horns.values:
+            satint = GeoInterpolator(
+                (longitude.values[:, :, horn], latitude.values[:, :, horn]),
+                (n_scan.values, n_subs),
+                (n_scan.values, np.arange(n_samples)),
+            )
+            lons_horn, lats_horn = satint.interpolate()
+            lons[:, :, horn] = lons_horn
+            lats[:, :, horn] = lats_horn
```
Is it possible to keep this dask compatible? (I don't know if geotiepoints allows it though)

mraspaud replied:
@BengtRydberg did you have on opinion on this?

BengtRydberg replied:
The tests passes if I provide dask arrays to geotiepoints functions, but it seems they run much slower, probably since geotiepoints uses numpy.

mraspaud commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -0,0 +1,441 @@
+        # This parameter is not applicable to ICI?
+        return None
+
+    @property
+    def cached_observation_azimuth(self):
+        """Get observation azimuth angles."""
+        observation_azimuth, _ = self.observation_azimuth_and_zenith
+        return observation_azimuth
+
+    @property
+    def cached_observation_zenith(self):
+        """Get observation zenith angles."""
+        _, observation_zenith = self.observation_azimuth_and_zenith
+        return observation_zenith
+
+    @property
+    def cached_solar_azimuth(self):
+        """Get solar azimuth angles."""
+        solar_azimuth, _ = self.solar_azimuth_and_zenith
+        return solar_azimuth
+
+    @property
+    def cached_solar_zenith(self):
+        """Get solar zenith angles."""
+        _, solar_zenith = self.solar_azimuth_and_zenith
+        return solar_zenith
+
+    @property
+    def cached_longitude(self):
+        """Get longitude coordinates."""
+        longitude, _ = self.longitude_and_latitude
+        return longitude
+
+    @property
+    def cached_latitude(self):
+        """Get latitude coordinates."""
+        _, latitude = self.longitude_and_latitude
+        return latitude
```
Why are these all called `cached_something`? I can't see any caching happening...

BengtRydberg replied:
That was bad naming so I removed the cache part from the name (it was the underlying functions that was using cache)

mraspaud commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -0,0 +1,441 @@
+        za.attrs = zenith.attrs
+        return aa, za
+
+    def _perform_calibration(self, variable, dataset_info):
```
Many functions here are called `perform_something`, could this be replace by a verb for clarity/brevity? (eg here: `_calibrate` would be enough.

BengtRydberg replied:
Thanks. I changed the function names according to your suggestion.

mraspaud commented on satpy/tests/reader_tests/test_ici_l1b_nc.py:
```diff
@@ -0,0 +1,432 @@
+TEST_FILE = 'test_file_ici_l1b_nc.nc'
+
+
+class TestIciL1bNCFileHandler(unittest.TestCase):
```
we're trying to move away from unittest and use pytest, so this class doesn't need to inherit from TestCase.
In that case, you can use the `tmp_path` fixture and remove the teardown.

BengtRydberg replied:
refactored so no using pytest

mraspaud commented on satpy/tests/reader_tests/test_ici_l1b_nc.py:
```diff
@@ -0,0 +1,432 @@
+            [198.22929022, 214.38700287, 230.54437184, 246.70146465],
+            [262.85833223, 279.01501371, 295.17153966, 311.32793429],
+        ])
+        assert np.allclose(bt, expected_bt)
```
`np.testing.assert_allclose` provide a nicer error message when failing.

mraspaud commented on satpy/tests/reader_tests/test_ici_l1b_nc.py:
```diff
@@ -0,0 +1,432 @@
+        mocked_calibrate.assert_not_called()
+        assert isinstance(dataset, xr.DataArray)
+
+    @patch('satpy.readers.ici_l1b_nc.IciL1bNCFileHandler._perform_orthorectification')  # noqa: E501
+    def test_get_dataset_orthorectifies_if_orthorect_data_defined(
+        self,
+        mocked_orthorectification,
+    ):
+        """Test get dataset orthorectifies if orthorect data is defined."""
+        dataset_id = {'name': 'lon_pixels_horn_1'}
+        dataset_info = {
+            'name': 'lon_pixels_horn_1',
+            'file_type': 'nc_ici_l1b_rad',
+            'file_key': 'cached_longitude',
+            'orthorect_data': 'data/navigation_data/delta_longitude',
+            'standard_name': 'longitude',
+            'n_horns': 0,
+            'modifiers': (),
+        }
+        self.reader.get_dataset(dataset_id, dataset_info)
+        mocked_orthorectification.assert_called_once_with(
+            ANY,  # the data
+            dataset_info['orthorect_data'],
+        )
+
+    @patch('satpy.readers.ici_l1b_nc.IciL1bNCFileHandler._perform_calibration')
+    def test_get_dataset_handles_calibration(
+        self,
+        mocked_calibration,
+    ):
+        """Test get dataset handles calibration."""
+        dataset_id = {'name': 'ici_1'}
+        dataset_info = {
+            'name': 'ici_1',
+            'file_type': 'nc_ici_l1b_rad',
+            'file_key': 'data/measurement_data/ici_radiance_183',
+            'coordinates': ['lat_pixels_horn_1', 'lon_pixels_horn_1'],
+            'n_183': 0,
+            'n_horns': 0,
+            'chan_index': 0,
```
Can we test that these functions (calibration, orthorectification) perform as expected instead of just testing that they are called?

BengtRydberg replied:
Changed so these tests are not mocked and the return value is now asserted.

mraspaud commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -0,0 +1,441 @@
+            'platform_name': self.spacecraft_name,
+        }
+
+        # Add a "quality_group" item to the dictionary with all the variables
+        # and attributes which are found in the 'quality' group of the product
+        quality_group = self['quality']
+        quality_dict = {}
+        for key in quality_group:
+            # Add the values (as Numpy array) of each variable in the group
+            # where possible
+            try:
+                quality_dict[key] = quality_group[key].values
+            except ValueError:
+                quality_dict[key] = None
+        # Add the attributes of the quality group
+        quality_dict.update(quality_group.attrs)
+        attributes['quality_group'] = quality_dict
+        return attributes
```
These comments indicate that the function could be made of smaller ones :)

BengtRydberg replied:
Agree :). Made smaller functions.

mraspaud replied:
thanks, very nice

mraspaud commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -0,0 +1,441 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2020 Satpy developers
```
```suggestion
# Copyright (c) 2022 Satpy developers
```

mraspaud reviewed (changes requested):
Thanks for adding this reader! I have a few comments inline.

BengtRydberg commented on AUTHORS.md:
```diff
@@ -78,3 +78,4 @@ The following people have made contributions to this project:
 - [praerien (praerien)](https://github.com/praerien)
 - [Xin Zhang (zxdawn)](https://github.com/zxdawn)
 - [Yufei Zhu (yufeizhu600)](https://github.com/yufeizhu600)
+- [Bengt Rydberg (BengtRydberg)](https://github.com/BengtRydberg)
```
fixed

BengtRydberg commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -0,0 +1,441 @@
+        self.orthorect = filetype_info.get('orthorect', True)
+
+    @cached_property
+    def observation_azimuth_and_zenith(self):
```
Thats true. I have refactored.

BengtRydberg commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -0,0 +1,441 @@
+        return end_time
+
+    @property
+    def spacecraft_name(self):
```
ok. changed to that.

BengtRydberg commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -0,0 +1,441 @@
+        za = np.degrees(np.arctan2(np.sqrt(x ** 2 + y ** 2), z))
+        aa.attrs = azimuth.attrs
+        za.attrs = zenith.attrs
+        return aa, za
```
thanks. yes, that seems to be unneeded so I removed that part.

BengtRydberg commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -0,0 +1,441 @@
+            b: temperature coefficient [K].
+
+        Returns:
+            numpy ndarray: array containing the calibrated brightness
+                temperature values.
```
yes, I updated the text.

BengtRydberg commented on satpy/tests/reader_tests/test_ici_l1b_nc.py:
```diff
@@ -0,0 +1,432 @@
+        """Test sub satellite path longitude."""
+        assert self.reader.ssp_lon is None
+
+    def test_perform_calibration_raises(self):
```
fixed

mraspaud commented on satpy/readers/ici_l1b_nc.py:
```diff
@@ -274,23 +242,67 @@ def _perform_viewing_angle_interpolation(
-        # transform from cartesian to spherical coords following Eumetsat spec
+        # transform to spherical coords with origin at north pole
+        # following Eumetsat spe
+        x, y, z = lonlat2xyz(aa, za)
```
I think this line and the following lines can be removed if you do `za = 90 - za` here, right? or am I misunderstanding this?

BengtRydberg replied:
You are right, this is not needed. Have removed it. Thanks!

mraspaud reviewed (commented):
Looks great, very nice with the fixtures and parametrizations!

Just one more comment here.

mraspaud commented:
@BengtRydberg Nice with the latest changes! I think it would be nice if you also added the appropriate identification keys in the yaml reader definitions, see for example what @adybbroe is doing here: https://github.com/pytroll/satpy/pull/2120

BengtRydberg commented:
> @BengtRydberg Nice with the latest changes! I think it would be nice if you also added the appropriate identification keys in the yaml reader definitions, see for example what @adybbroe is doing here: #2120

Agree. Added the identification keys.

mraspaud reviewed (approved):
LGTM, thanks a lot for adding support for ICI. @adybbroe could you also have a look at this PR since you been working with this kind of instruments?

adybbroe commented on satpy/etc/readers/ici_l1b_nc.yaml:
```diff
@@ -0,0 +1,650 @@
+    name:
+      required: true
+    frequency_double_sideband:
+      type: !!python/name:satpy.readers.aapp_mhs_amsub_l1c.FrequencyDoubleSideBand
```
We eventually has to adapt this once the other PR #2120 is merged. I introduced a common MW module not specific to AMSU/MHS

adybbroe reviewed (commented):
LGTM. Very nice work! 
Just one thing: Can you add the reader to the list of supported readers in the documentation?
See the `index.rst` file

Also a question, does the order of the coordinates matter? I have the lons first and the lats second, I see you have it the other way around. Have you tested reading, resampling and displaying one of the channels?

djhoese commented:
You aren't wrong about the coordinates being sorted by standard name, but I'm also not sure I like the inconsistency between these readers and others.

BengtRydberg commented:
> You aren't wrong about the coordinates being sorted by standard name, but I'm also not sure I like the inconsistency between these readers and others.

Agree. I have changed the order. I was just not sure what was the "standard order" but I can see that this is how it is done in the yaml file for most of the readers in the repo.

gerritholl commented:
Could you merge main and resolve the merge conflicts?  I think merging main will also take care of the SEVIRI test failures.

BengtRydberg commented:
> Could you merge main and resolve the merge conflicts? I think merging main will also take care of the SEVIRI test failures.

Yes, I have merged main.

mraspaud reviewed (approved):
LGTM! Thanks for putting the effort and provide a clean PR
